google/gax/api_callable.py

```diff
@@ -262,63 +265,42 @@ class ApiCallable(object):
-                `defaults`, a runtime error will result at call time. If the
-                `max_attempts` setting cannot be determined either from
-                `options` or `defaults`, a runtime error will result at call
-                time.
-            is_idempotent: If set, this call is marked as idempotent. Idempotent
-                calls' default retrying behavior may be set in `defaults`.
+            settings: A CallSettings object from which the settings for this
+                call are drawn.
 
         Returns:
             An ApiCallable object.
         """
-        self.options = CallOptions() if options is None else options
-        self.defaults = defaults
         self.func = func
-        self.is_idempotent = is_idempotent
-
-    def _call_settings(self):
-        self.options.normalize()
-        if self.is_idempotent and self.options.is_retrying is None:
-            is_retrying = self.defaults.is_idempotent_retrying
-        else:
-            is_retrying = self.options.is_retrying
-        page_descriptor = self.options.page_streaming
-        max_attempts = self.options.max_attempts
-        timeout = self.options.timeout
-        if self.defaults is not None:
-            if max_attempts is None:
-                max_attempts = self.defaults.max_attempts
-            if timeout is None:
-                timeout = self.defaults.timeout
-        return is_retrying, max_attempts, page_descriptor, timeout
+        self.settings = settings
 
     def __call__(self, *args, **kwargs):
         the_func = self.func
-        is_retrying, max_attempts, page_desc, timeout = self._call_settings()
 
         # Update the_func using each of the applicable function decorators
         # before calling.
-        if is_retrying:
-            the_func = _retryable(the_func, max_attempts)
-        if page_desc:
+        if self.settings.is_retrying:
+            the_func = _retryable(the_func, self.settings.max_attempts)
+        if self.settings.page_descriptor:
+            if self.settings.bundler and self.settings.bundle_descriptor:
+                raise ValueError('ApiCallable has incompatible settings: '
+                                 'bundling and page streaming')
             the_func = _page_streamable(
                 the_func,
-                page_desc.request_page_token_field,
-                page_desc.response_page_token_field,
-                page_desc.resource_field,
-                timeout)
+                self.settings.page_descriptor.request_page_token_field,
+                self.settings.page_descriptor.response_page_token_field,
+                self.settings.page_descriptor.resource_field,
+                self.settings.timeout)
         else:
-            the_func = _add_timeout_arg(the_func, timeout)
+            the_func = _add_timeout_arg(the_func, self.settings.timeout)
+            if self.settings.bundler and self.settings.bundle_descriptor:
+                the_func = _bundleable(
```

Comment:
N.B: At this point, _bundleable is called with the_func which is still a callable([[request], response])
I refer to this in another comment.